src/flow/src/batching_mode/state.rs

```diff
@@ -0,0 +1,375 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Batching mode task state, which changes frequently
+
+use std::collections::BTreeMap;
+use std::time::Duration;
+
+use common_telemetry::debug;
+use common_telemetry::tracing::warn;
+use common_time::Timestamp;
+use datatypes::value::Value;
+use session::context::QueryContextRef;
+use snafu::ResultExt;
+use tokio::sync::oneshot;
+use tokio::time::Instant;
+
+use crate::adapter::FlowId;
+use crate::batching_mode::task::BatchingTask;
+use crate::batching_mode::MIN_REFRESH_DURATION;
+use crate::error::{DatatypesSnafu, InternalSnafu, TimeSnafu};
+use crate::Error;
+
+#[derive(Debug)]
+pub struct TaskState {
+    /// Query context
+    pub(crate) query_ctx: QueryContextRef,
+    /// last query complete time
+    last_update_time: Instant,
+    /// last time query duration
+    last_query_duration: Duration,
+    /// Dirty Time windows need to be updated
+    /// mapping of `start -> end` and non-overlapping
+    pub(crate) dirty_time_windows: DirtyTimeWindows,
+    exec_state: ExecState,
+    /// Shutdown receiver
+    pub(crate) shutdown_rx: oneshot::Receiver<()>,
+}
+impl TaskState {
+    pub fn new(query_ctx: QueryContextRef, shutdown_rx: oneshot::Receiver<()>) -> Self {
+        Self {
+            query_ctx,
+            last_update_time: Instant::now(),
+            last_query_duration: Duration::from_secs(0),
+            dirty_time_windows: Default::default(),
+            exec_state: ExecState::Idle,
+            shutdown_rx,
+        }
+    }
+
+    /// called after last query is done
+    /// `is_succ` indicate whether the last query is successful
+    pub fn after_query_exec(&mut self, elapsed: Duration, _is_succ: bool) {
+        self.exec_state = ExecState::Idle;
+        self.last_query_duration = elapsed;
+        self.last_update_time = Instant::now();
+    }
+
+    /// wait for at least `last_query_duration`, at most `max_timeout` to start next query
+    pub fn get_next_start_query_time(&self, max_timeout: Option<Duration>) -> Instant {
+        let next_duration = max_timeout
+            .unwrap_or(self.last_query_duration)
+            .min(self.last_query_duration);
+        let next_duration = next_duration.max(MIN_REFRESH_DURATION);
+
+        self.last_update_time + next_duration
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct DirtyTimeWindows {
+    /// windows's `start -> end` and non-overlapping
+    /// `end` is exclusive(and optional)
+    windows: BTreeMap<Timestamp, Option<Timestamp>>,
```

Comment:
non-blocking: using a sorted Vec can be more effective.

Reply:
> non-blocking: using a sorted Vec can be more effective.

need to insert at random position so need a ordered map